connectors/src/connectors/notion/lib/parents.ts

```diff
@@ -0,0 +1,142 @@
+ * Doc](https://www.notion.so/dust-tt/Engineering-e0f834b5be5a43569baaf76e9c41adf2?p=3d26536a4e0a464eae0c3f8f27a7af97&pm=s)
+ * and the field documentation [in core]() for relevant details
+ */
+async function _getParents(
+  dataSourceInfo: DataSourceInfo,
+  document: {
+    notionId: string;
+    parentType: string | null | undefined;
+    parentId: string | null | undefined;
+  }
+): Promise<string[]> {
+  const parents: string[] = [document.notionId];
+  switch (document.parentType) {
+    case "workspace":
+      return parents;
+    case "block":
+      // rare cases in which doing something here is useful
+      // are ignored for now, see the design doc for details
+      return parents;
+    case "page":
+    case "database": {
+      // retrieve the parent from notion connectors db
+      // and add it to the parents array
+      const parent = await getNotionPageFromConnectorsDb(
+        dataSourceInfo,
+        document.parentId as string // (cannot be null here)
+      );
+      if (!parent) {
+        // The parent is either not synced yet (not an issue, see design doc) or
+        // is not in Dust's scope, in both cases we can just return the page id
+        return parents;
+      }
+      return parents.concat(
+        await getParents(dataSourceInfo, {
+          notionId: parent.notionPageId,
+          parentType: parent.parentType,
+          parentId: parent.parentId,
+        })
+      );
+    }
+    default:
+      throw new Error(`Unhandled parent type ${document.parentType}`);
+  }
+}
+
+export const getParents = memoize(_getParents, (dataSourceInfo, document) => {
+  return `${dataSourceInfo.dataSourceName}:${document.notionId}`;
+});
+
+export async function updateAllParentsFields(
+  dataSourceConfig: DataSourceConfig,
+  documents: (NotionPage | NotionDatabase)[]
+) {
+  /* Computing all descendants, then updating, ensures the field is updated only
+    once per page, limiting the load on the Datasource */
+  const pagesToUpdate = await getPagesToUpdate(documents, dataSourceConfig);
+
+  // Update everybody's parents field
+  for (const page of pagesToUpdate) {
+    const parents = await getParents(dataSourceConfig, {
+      notionId: page.notionPageId,
+      parentType: page.parentType,
+      parentId: page.parentId,
+    });
+
+    await updateDocumentParentsField(
+      dataSourceConfig,
+      `notion-${page.notionPageId}`,
+      parents
+    );
+  }
+}
+
+/**  Get ids of all pages whose parents field should be updated: inital pages in
+ * documentIds, and all the descendants of documentIds that are pages (including
+ * children of databases)
+ *
+ * Note: databases are not stored in the Datasource, so they don't need to be
+ * updated
+ */
+async function getPagesToUpdate(
+  documents: (NotionPage | NotionDatabase)[],
+  dataSourceConfig: DataSourceConfig
+): Promise<NotionPage[]> {
+  const pagesToUpdate: NotionPage[] = [];
+
+  let i = 0;
+  while (i < documents.length) {
```

Comment:
Are you sure documents.length dynamically grow here?

Reply:
Thanks for the catch !

Reply:
Yes, hopefully it behaves as expected by evaluating the condition dynamically at each run of the loop, documents.length is not replaced with its value at first exec (length being the property of an object)

Reply:
I find it slightly confusing